light-clients/ics10-grandpa-cw/src/contract.rs

```diff
@@ -0,0 +1,361 @@
+	ics02_client::{
+		client_def::{ClientDef, ConsensusUpdateResult},
+		context::{ClientKeeper, ClientReader},
+		height::Height,
+	},
+	ics24_host::identifier::ClientId,
+};
+use ics08_wasm::client_state::ClientState as WasmClientState;
+use ics10_grandpa::{
+	client_def::GrandpaClient,
+	client_message::{ClientMessage, RelayChainHeader},
+	consensus_state::ConsensusState,
+};
+use light_client_common::{verify_membership, verify_non_membership};
+use prost::Message;
+use sp_core::H256;
+use sp_runtime::traits::{BlakeTwo256, Header};
+use sp_runtime_interface::unpack_ptr_and_len;
+use std::{collections::BTreeSet, str::FromStr};
+/*
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:ics10-grandpa-cw";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+*/
+
+pub const CHANNELS_CONNECTION: Map<Bytes, Vec<(Bytes, Bytes)>> = Map::new("channels_connection");
+pub const CLIENT_UPDATE_TIME: Map<(Bytes, Bytes), u64> = Map::new("client_update_time");
+pub const CLIENT_UPDATE_HEIGHT: Map<(Bytes, Bytes), Bytes> = Map::new("client_update_height");
+pub const CHANNEL_COUNTER: Item<u32> = Item::new("channel_counter");
+pub const EXPECTED_BLOCK_TIME: Item<u64> = Item::new("expected_block_time");
+pub const CONNECTION_PREFIX: Item<Vec<u8>> = Item::new("connection_prefix");
+pub const CONNECTION_COUNTER: Item<u32> = Item::new("connection_counter");
+pub const CLIENT_COUNTER: Item<u32> = Item::new("client_counter");
+pub const CODE_ID: Item<Vec<u8>> = Item::new("code_id");
+pub const HOST_CONSENSUS_STATE: Map<u64, ConsensusState> = Map::new("host_consensus_state");
+pub const CONSENSUS_STATES_HEIGHTS: Map<Bytes, BTreeSet<Height>> =
+	Map::new("consensus_states_heights");
+pub const GRANDPA_HEADER_HASHES_STORAGE: Item<Vec<H256>> = Item::new("grandpa_header_hashes");
+pub const GRANDPA_HEADER_HASHES_SET_STORAGE: Map<Vec<u8>, ()> =
+	Map::new("grandpa_header_hashes_set");
+
+pub const GRANDPA_BLOCK_HASHES_CACHE_SIZE: usize = 500;
+
+#[derive(Clone, Copy, Debug, PartialEq, Default, Eq)]
+pub struct HostFunctions;
+
+impl light_client_common::HostFunctions for HostFunctions {
+	type BlakeTwo256 = BlakeTwo256;
+}
+
+impl grandpa_light_client_primitives::HostFunctions for HostFunctions {
+	type Header = RelayChainHeader;
+
+	fn ed25519_verify(
+		sig: &sp_core::ed25519::Signature,
+		msg: &[u8],
+		pub_key: &sp_core::ed25519::Public,
+	) -> bool {
+		use ed25519_zebra::{Signature, VerificationKey as PublicKey, VerificationKeyBytes};
+		let bytes: [u8; 64] = (sig.clone()).try_into().expect("signature is 64 bytes; qed");
+		let sig = Signature::from(bytes);
+		let slice: &[u8] = pub_key.as_ref();
+		let pub_key = PublicKey::try_from(
+			VerificationKeyBytes::try_from(slice).expect("key is always valid; qed"),
+		)
+		.expect("key is always valid; qed");
+		pub_key.verify(&sig, msg).is_ok()
+	}
+
+	fn insert_relay_header_hashes(_headers: &[<Self::Header as Header>::Hash]) {
+		// implementation of this method is in `Context`
```

Comment:
shouldn't it panic then?

Reply:
No, we can't, because this method is always called